Cmdline/Action/Upgrade.cs

```diff
@@ -160,5 +159,80 @@ public int RunCommand(CKAN.GameInstance ksp, object raw_options)
 
             return Exit.OK;
         }
+
+        /// <summary>
+        /// Upgrade some modules by their CkanModules
+        /// </summary>
+        /// <param name="manager">Game instance manager to use</param>
+        /// <param name="user">IUser object for output</param>
+        /// <param name="ksp">Game instance to use</param>
+        /// <param name="modules">List of modules to upgrade</param>
+        public static void UpgradeModules(GameInstanceManager manager, IUser user, CKAN.GameInstance ksp, bool ConfirmPrompt, List<CkanModule> modules)
+        {
+            UpgradeModules(manager, user, ksp,
+                (ModuleInstaller installer, NetAsyncModulesDownloader downloader, RegistryManager regMgr, ref HashSet<string> possibleConfigOnlyDirs) =>
+                    installer.Upgrade(modules, downloader,
+                        ref possibleConfigOnlyDirs, regMgr, true, true, ConfirmPrompt),
+                m => modules.Add(m)
+            );
+        }
+
+        /// <summary>
+        /// Upgrade some modules by their CkanModules
+        /// </summary>
+        /// <param name="manager">Game instance manager to use</param>
+        /// <param name="user">IUser object for output</param>
+        /// <param name="ksp">Game instance to use</param>
+        /// <param name="identsAndVersions">List of identifier[=version] to upgrade</param>
+        public static void UpgradeModules(GameInstanceManager manager, IUser user, CKAN.GameInstance ksp, List<string> identsAndVersions)
+        {
+            UpgradeModules(manager, user, ksp,
+                (ModuleInstaller installer, NetAsyncModulesDownloader downloader, RegistryManager regMgr, ref HashSet<string> possibleConfigOnlyDirs) =>
+                    installer.Upgrade(identsAndVersions, downloader,
+                        ref possibleConfigOnlyDirs, regMgr, true),
+                m => identsAndVersions.Add(m.identifier)
+            );
+        }
+
+        // System.Action<ref T> isn't allowed
+        private delegate void AttemptUpgradeAction(ModuleInstaller installer, NetAsyncModulesDownloader downloader, RegistryManager regMgr, ref HashSet<string> possibleConfigOnlyDirs);
+
+        private static void UpgradeModules(
```

Comment:
I think this one needs some documenation.
It takes a lot of scrolling back and forth to figure out what the callbacks do and why they're there.
Something that explains the this private method provides the outer logic to reduce code duplication,
the `attemptUpgradeCallback` should call the according `ModuleInstaller.Upgrade()`,
that `addUserChoiceCallback` then should modify the list of mods to install to add the user's choice after a `TooManyModsProvideKraken` was thrown, etc.

Reply:
Does that help?  1bca7e8d0fdc0882c6f51dc19e78be3d826ac90e

Reply:
Yup, looks good. Thanks!